File: 3D Touch/ViewController.m
//
//  ViewController.m
//  3D Touch
//
//  Created by Jay Versluis on 21/09/2015.
//  Copyright © 2015 Pinkstone Pictures LLC. All rights reserved.
//

#import "ViewController.h"
#import "PreviewViewController.h"

@interface ViewController () <UIViewControllerPreviewingDelegate>

@property (nonatomic, strong) UILongPressGestureRecognizer *longPress;

@end

@implementation ViewController

- (void)viewDidLoad {
    
    [super viewDidLoad];
}

- (void)viewWillAppear:(BOOL)animated {
    
    [self check3DTouch];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)check3DTouch {
    
    // register for 3D Touch (if available)
    if (self.traitCollection.forceTouchCapability == UIForceTouchCapabilityAvailable) {
        
        [self registerForPreviewingWithDelegate:(id)self sourceView:self.view];
        NSLog(@"3D Touch is available! Hurra!");
        
        // no need for our alternative anymore
        self.longPress.enabled = NO;
        
    } else {
        
        NSLog(@"3D Touch is not available on this device. Sniff!");
        
        // handle a 3D Touch alternative (long gesture recognizer)
        self.longPress.enabled = YES;
        
        }
}

- (UILongPressGestureRecognizer *)longPress {
    
    if (!_longPress) {
        _longPress = [[UILongPressGestureRecognizer alloc]initWithTarget:self action:@selector(showPeek)];
        [self.view addGestureRecognizer:_longPress];
    }
    return _longPress;
}


# pragma mark - 3D Touch Delegate

- (UIViewController *)previewingContext:(id<UIViewControllerPreviewing>)previewingContext viewControllerForLocation:(CGPoint)location {
    
    // check if we're not already displaying a preview controller
    if ([self.presentedViewController isKindOfClass:[PreviewViewController class]]) {
        return nil;
    }
    
    // shallow press: return the preview controller here (peek)
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    UIViewController *previewController = [storyboard instantiateViewControllerWithIdentifier:@"PreviewView"];
    
    return previewController;
}

- (void)previewingContext:(id<UIViewControllerPreviewing>)previewingContext commitViewController:(UIViewController *)viewControllerToCommit {
    
    // deep press: bring up the commit view controller (pop)
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    UIViewController *commitController = [storyboard instantiateViewControllerWithIdentifier:@"CommitView"];
    
    [self showViewController:commitController sender:self];
    
    // alternatively, use the view controller that's being provided here (viewControllerToCommit)
}

- (void)traitCollectionDidChange:(UITraitCollection *)previousTraitCollection {
    
    // called when the interface environment changes
    // one of those occasions would be if the user enables/disables 3D Touch
    // so we'll simply check again at this point
    
    [self check3DTouch];
}


#pragma mark - 3D Touch Alternative

- (void)showPeek {
    
    // disable gesture so it's not called multiple times
    self.longPress.enabled = NO;
    
    // present the preview view controller (peek)
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    PreviewViewController *preview = [storyboard instantiateViewControllerWithIdentifier:@"PreviewView"];
    
    UIViewController *presenter = [self grabTopViewController];
    [presenter showViewController:preview sender:self];
    
}

- (UIViewController *)grabTopViewController {
    
    // helper method to always give the top most view controller
    // avoids "view is not in the window hierarchy" error
    // http://stackoverflow.com/questions/26022756/warning-attempt-to-present-on-whose-view-is-not-in-the-window-hierarchy-sw
    
    UIViewController *top = [UIApplication sharedApplication].keyWindow.rootViewController;
    
    while (top.presentedViewController) {
        top = top.presentedViewController;
    }
    
    return top;
}

@end

File: 3D Touch - Swift/3D Touch Swift/ViewController.swift
//
//  ViewController.swift
//  3D Touch Swift
//
//  Created by Jay Versluis on 04/12/2015.
//  Copyright © 2015 Pinkstone Pictures LLC. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UIViewControllerPreviewingDelegate {
    
    lazy var longPress: UILongPressGestureRecognizer = {
        let temp = UILongPressGestureRecognizer.init(target: self, action: #selector(ViewController.showPeek))
        // self.view.addGestureRecognizer(longPress)
        return temp
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.addGestureRecognizer(longPress)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewWillAppear(_ animated: Bool) {
        
        // check if 3D Touch is available
        self.check3DTouch()
    }
    
    func check3DTouch() {
        
        // register for 3D Touch (if available)
        if self.traitCollection.forceTouchCapability == UIForceTouchCapability.available {
            
            self.registerForPreviewing(with: self, sourceView: self.view)
            print("3D Touch is available. Excellent!")
            
            // and disable the long press gesture
            self.longPress.isEnabled = false

        } else {
            
            print("3D Touch is not available. Dang!")
            
            // handle alternative long press recogniser
            self.longPress.isEnabled = true
            
        }
    }

    // MARK: 3D Touch Delegate
    
    func previewingContext(_ previewingContext: UIViewControllerPreviewing, viewControllerForLocation location: CGPoint) -> UIViewController? {
        
        // check if we're not already showing a preview
        if ((self.presentedViewController?.isKind(of: PreviewViewController.self)) != nil) {
            return nil
        }
        
        // PEEK (shallow press): return the preview view controller here
        let storyboard = UIStoryboard.init(name: "Main", bundle: nil)
        let previewView = storyboard.instantiateViewController(withIdentifier: "PreviewView")
        
        return previewView
    }
    
    func previewingContext(_ previewingContext: UIViewControllerPreviewing, commit viewControllerToCommit: UIViewController) {
        
        // POP (deep press): return the commit view controller here
        let storyboard = UIStoryboard.init(name: "Main", bundle: nil)
        let commitView = storyboard.instantiateViewController(withIdentifier: "CommitView")
        
        self.show(commitView, sender: self)
        
        // alternatively, use the view controller that's being provided here (viewControllerToCommit)
    }
    
    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        
        // called when the interface environment changes
        // one of those occasions would be if the user enables/disables 3D Touch
        // so we'll simply check again at this point
        self.check3DTouch()
    }
    
    // MARK: 3D Touch Alternative
    
    @objc func showPeek() {
        
        print("showing peek")
        
        // disable long press gesture so it's not called mutiple times
        self.longPress.isEnabled = false
        
        // present the preview (peek)
        let storyboard = UIStoryboard.init(name: "Main", bundle: nil)
        let previewView = storyboard.instantiateViewController(withIdentifier: "PreviewView")
        
        let presenter = self.grabTopViewController()
        presenter.show(previewView, sender: self)
        
    }
    
    func grabTopViewController() -> UIViewController {
        
        // helper method to always give the top most view controller
        // avoids "view is not in the window hierarchy" error
        // http://stackoverflow.com/questions/26022756/warning-attempt-to-present-on-whose-view-is-not-in-the-window-hierarchy-sw
        
        var top = UIApplication.shared.keyWindow?.rootViewController
        while ((top?.presentedViewController) != nil) {
            top = top?.presentedViewController
        }
        
        return top!
    }

}

